api/desecapi/serializers.py

```diff
@@ -366,22 +377,56 @@ def to_internal_value(self, data):
             if not isinstance(item, dict):
                 self.fail('invalid', datatype=type(item).__name__)
             key = self._key(item)
-            items = indices.setdefault(key[0], {}).setdefault(key[1], set())
+            if key[0] not in indices:
+                types = self.child.domain.rrset_set.filter(subname=key[0]).values_list('type', flat=True)
+                indices[key[0]] = {type_: {None} for type_ in types}
+            items = indices[key[0]].setdefault(key[1], set())
             items.add(idx)
 
+        collapsed_indices = copy.deepcopy(indices)
+        for idx, item in enumerate(data):
+            if item.get('records') == []:
+                key = self._key(item)
+                collapsed_indices[key[0]][key[1]] -= {idx, None}
+
         # Iterate over all rows in the data given
         for idx, item in enumerate(data):
             try:
                 # see if other rows have the same key
                 key = self._key(item)
-                if len(indices[key[0]][key[1]]) > 1:
+                data_indices = indices[key[0]][key[1]] - {None}
+                if len(data_indices) > 1:
                     raise serializers.ValidationError({
                         'non_field_errors': [
                             'Same subname and type as in position(s) %s, but must be unique.' %
-                            ', '.join(map(str, indices[key[0]][key[1]] - {idx}))
+                            ', '.join(map(str, data_indices - {idx}))
                         ]
                     })
 
+                # see if other rows violate CNAME exclusivity
+                if item.get('records') != []:
+                    conflicting_indices_by_type = {k: v for k, v in collapsed_indices[key[0]].items()
+                                                   if (k == 'CNAME') != (key[1] == 'CNAME')}
+                    if any(conflicting_indices_by_type.values()):
+                        types_by_position = {}
+                        for type_, conflict_positions in conflicting_indices_by_type.items():
+                            for position in conflict_positions:
+                                types_by_position.setdefault(position, []).append(type_)
+                        # Sort by position, None at the end
+                        types_by_position = dict(sorted(types_by_position.items(), key=lambda x: (x[0] is None, x)))
+                        db_conflicts = types_by_position.pop(None, None)
+                        if db_conflicts: types_by_position['database'] = db_conflicts
+                        for position, types in types_by_position.items():
+                            types_by_position[position] = ', '.join(types)
+                        types_by_position = [f'{position} ({types})' for position, types in types_by_position.items()]
+                        types_by_position = ', '.join(types_by_position)
```

Comment:
maybe use [`pprint.pformat`](https://docs.python.org/3/library/pprint.html#pprint.pformat) to save some lines here

Reply:
doesn't really work, will have extra apostrophes etc.